tsl/src/fdw/data_node_chunk_assignment.c

```diff
@@ -75,32 +58,48 @@ DataNodeChunkAssignment *
 data_node_chunk_assignment_assign_chunk(DataNodeChunkAssignments *scas, RelOptInfo *chunkrel)
 {
 	DataNodeChunkAssignment *sca = get_or_create_sca(scas, chunkrel->serverid, NULL);
-	RangeTblEntry *rte = planner_rt_fetch(chunkrel->relid, scas->root);
+	TsFdwRelInfo *chunk_private = fdw_relinfo_get(chunkrel);
 	MemoryContext old;
 
 	/* Should never assign the same chunk twice */
 	Assert(!bms_is_member(chunkrel->relid, sca->chunk_relids));
 
-	old = MemoryContextSwitchTo(scas->mctx);
-
 	/* If this is the first chunk we assign to this data node, increment the
 	 * number of data nodes with one or more chunks on them */
-	if (list_length(sca->chunk_oids) == 0)
+	if (list_length(sca->chunks) == 0)
 		scas->num_nodes_with_chunks++;
 
+	scas->total_num_chunks++;
+
+	/*
+	 * Use the cached ChunkDataNode data to find the relid of the chunk on the
+	 * data node.
+	 */
+	int remote_chunk_relid = InvalidOid;
+	ListCell *lc;
+	foreach (lc, chunk_private->chunk->data_nodes)
```

Comment:
Maybe move this to separate function?

Reply:
Which part, the loop only? Too small, let's just keep it inline. It's just boilerplate putting data from one place to another, I feel that splitting it into small functions just makes it harder to read.